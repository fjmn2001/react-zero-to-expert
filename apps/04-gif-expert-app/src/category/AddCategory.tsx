import {
  ChangeEvent,
  Dispatch,
  FormEvent,
  SetStateAction,
  useState,
} from "react"

const AddCategory = ({
  setCategories,
}: {
  setCategories: Dispatch<SetStateAction<string[]>>
}) => {
  const [inputValue, setInputValue] = useState("")

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.currentTarget.value)
  }

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault()

    if (inputValue.trim().length === 0) {
      return
    }

    setCategories((categories) => [inputValue, ...categories])
    setInputValue("")
  }

  return (
    <form onSubmit={handleSubmit}>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  )
}

export default AddCategory
