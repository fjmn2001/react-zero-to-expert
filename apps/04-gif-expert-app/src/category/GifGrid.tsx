const apiKey = "g13oXIzvbtzxNnCw6hjWVR31UjL0yGVw"

const GifGrid = ({ category }: { category: string }) => {
  const getGifs = async () => {
    const query = "Rick"
    const url = `https://api.giphy.com/v1/gifs/search?api_key=${apiKey}&q=${query}&limit=10`
    const response = await fetch(url)
    const { data } = await response.json()

    const gifs = data.map(
      ({
        id,
        title,
        images,
      }: {
        id: string
        title: string
        // eslint-disable-next-line camelcase
        images: { downsized_large: { url: string } }
      }) => {
        return { id, title, url: images.downsized_large.url }
      }
    )

    console.log(gifs)
  }

  // id
  // downsize_large
  getGifs()

  return (
    <>
      <h3>Gif Grid</h3>
      <h4>{category}</h4>
    </>
  )
}

export default GifGrid
