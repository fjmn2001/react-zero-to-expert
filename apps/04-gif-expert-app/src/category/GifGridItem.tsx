import { Image } from "./types"

const GifGridItem = ({ image }: { image: Image }) => {
  const { title, url } = image

  return (
    <div className={"card animate__animated animate__fadeIn"}>
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  )
}

export default GifGridItem
