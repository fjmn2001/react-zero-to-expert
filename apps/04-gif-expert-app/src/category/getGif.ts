const apiKey = "g13oXIzvbtzxNnCw6hjWVR31UjL0yGVw"

export const getGifs = async (category: string) => {
  const url = `https://api.giphy.com/v1/gifs/search?api_key=${apiKey}&q=${encodeURI(
    category
  )}&limit=10`
  const response = await fetch(url)
  const { data } = await response.json()

  return data.map(
    ({
      id,
      title,
      images,
    }: {
      id: string
      title: string
      // eslint-disable-next-line camelcase
      images: { downsized_large: { url: string } }
    }) => {
      return { id, title, url: images.downsized_large.url }
    }
  )
}
