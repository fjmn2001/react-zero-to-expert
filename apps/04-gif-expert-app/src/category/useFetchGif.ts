import { useEffect, useState } from "react"
import { getGifs } from "./getGif"
import { Image } from "./types"

export const useFetchGif = (category: string) => {
  const [state, setState] = useState({
    data: [],
    loading: true,
  })

  useEffect(() => {
    getGifs(category).then((images) =>
      setState({
        data: images,
        loading: false,
      })
    )
  }, [category])

  return state as { data: Image[]; loading: boolean }
}
