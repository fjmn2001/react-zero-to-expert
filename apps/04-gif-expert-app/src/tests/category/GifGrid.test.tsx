import { beforeEach, describe, expect, MockedFunction, test, vi } from "vitest"
import { shallow } from "enzyme"
import GifGrid from "../../category/GifGrid"
import { useFetchGif } from "../../category/useFetchGif"
import { Image } from "../../category/types"

vi.mock("../../category/useFetchGif")

describe("Testing <GifGrid/>", () => {
  const category = "One Punch"

  beforeEach(() => {
    vi.resetAllMocks()
  })

  test("it should show correctly", () => {
    ;(useFetchGif as MockedFunction<any>).mockReturnValue({
      data: [],
      loading: true,
    })
    const wrapper = shallow(<GifGrid category={category} />)
    expect(wrapper).toMatchSnapshot()
  })

  test("it should show items when images have been loaded", () => {
    const gifs: Image[] = [
      {
        id: "custom-id",
        title: "custom-title",
        url: "custom-url",
      },
    ]
    ;(useFetchGif as MockedFunction<any>).mockReturnValue({
      data: gifs,
      loading: false,
    })
    const wrapper = shallow(<GifGrid category={category} />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find("p").exists()).toBe(false)
    expect(wrapper.find("GifGridItem").length).toBe(1)
  })
})
