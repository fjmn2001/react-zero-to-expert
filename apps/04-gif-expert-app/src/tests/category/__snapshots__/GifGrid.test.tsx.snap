// Vitest Snapshot v1

exports[`Testing <GifGrid/> > it should show correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  >
    <ol />
  </div>
</Fragment>
`;

exports[`Testing <GifGrid/> > it should show items when images have been loaded 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <ol>
      <GifGridItem
        image={
          {
            "id": "custom-id",
            "title": "custom-title",
            "url": "custom-url",
          }
        }
        key="custom-id"
      />
    </ol>
  </div>
</Fragment>
`;
